### ISSUE TITLE
Emacs blocks on flycheck-rust-find-target

### ISSUE THREAD

**Issue by Pireax** [flycheck/flycheck-rust#44]
Hello,
When using this function directly or automatically through flycheck it blocks up the UI for a few seconds. Is it possible to make this not block?

**Comment by lunaryorn** [comment]
@Pireax Yes, it'd be possible, but only with a major refactoring.  To me personally it's not worth the effort, to be honest, since the function should only run once per buffer, and I'm not doing much Rust recently anyway.

Maybe someone from @flycheck/rust can take a look, but we would really appreciate if you could try and open a pull request.

**Comment by fmdkdd** [comment]
Well, a few *seconds* seems quite bad indeed.  I haven't encountered these kind of durations.  I imagine that in some setups, executing `cargo read-manifest` and parsing the JSON can take its toll.

I guess we could call cargo asynchronously (is that what you had in mind @lunaryorn ?), but we would still have to parse the JSON synchronously.

Ultimately though, my hope is that I find time to interface flycheck with the Rust Language Server, and let the server deal with all this.

In the meantime, you can use a `.dir-locals.el` file in your Rust project to set the cargo targets directly and disable the flycheck-rust-setup hook.

**Comment by lunaryorn** [comment]
@fmdkdd Yes, that's what I mean: We could call cargo asynchronously and set the variables in the process handler once cargo is finished.  But I don't know whether it's worth the effort.

**Comment by Pireax** [comment]
Would something along the lines of [https://gist.github.com/Pireax/7b3c58a3afd57d2f6b37ae9c0abf394b](this) work without having any unwanted consequences?

**Comment by lunaryorn** [comment]
@Pireax Could you please link to a diff?

**Comment by Pireax** [comment]
[Here you go.](https://github.com/Pireax/flycheck-rust/commit/5d5d9e7f1a33f5a6b5b3f19d077cc15e8939f118)

**Comment by lunaryorn** [comment]
@Pireax I don't think that'd work.  

"async" spawns a separate Emacs instance, which doesn't inherit any values or libraries from the current sessions automatically.   I presume this `let`-bind of `load-path` is to propagate the value of `load-path` to the asynchronous function?  In this case we're—from a cursory look—at least missing `exec-path` and `flycheck-executable-find` as well; both of these are relevant in `flycheck-rust-find-target`.

It's also not the most efficient solution, to be honest: It's spawning an entire new Emacs session and load all of `flycheck` and `flycheck-rust` a second time just to spawn another process—which Emacs can already do asynchronously with `start-process`.

I would want to see `start-process` being used to call out to `cargo` first, to see whether that's the problem.  If it turns out that parsing the JSON output is slow, too, we can bring async.el in, but then confined to just the parsing code, so to avoid having to pass any variables down to the async function, let alone loading the entire Flycheck library twice.

**Comment by fmdkdd** [comment]
Also, wouldn't there be potential sequencing issues?  If the variables are set asynchronously, presumably after a few seconds, well after the `flycheck-mode-hook is run`.  I guess at least you wouldn't disable the `rust` checker like you would have with the synchronized version.

**Comment by Pireax** [comment]
Just setting the exec-path like I did with load-path would work right? And indeed it is a bit of a roundabout way to do it but you don't really notice the performance as it is done "behind the scenes". Anyways I'll benchmark the flycheck-rust-find-target function and look if the JSON parsing is the problem or if its the cargo call.

**Comment by Pireax** [comment]
It's definitely only call-process that is causing this issue, it takes 6 seconds for me... The whole function takes 6.5 seconds.

**Comment by lunaryorn** [comment]
@Pireax In this case I'd rather prefer to go without an additional dependency, and call cargo asynchronously with `start-process`.

**Comment by Pireax** [comment]
Starting a new emacs process using `start-process` definetly is a lot faster than a call to `start-process` with cargo(not taking into account how long it takes for the process to finish). I don't know if this is the case just on my end but if it isn't I think it would make sense to use emacs-async.

**Comment by lunaryorn** [comment]
@Pireax `start-process` does not block… I guess you referred to `call-process` which is what we're using currently.  But it makes no sense to `start-process` a separate Emacs instance with an additional dependency just to `call-process` another program: You could just `start-process` the other program—`cargo` in this case—directly.

IOW, if it's the cargo invocation that is slow we have no need for `async.el`—Emacs can run asynchronous processes out of the box.

**Comment by Pireax** [comment]
I am talking about `start-process`, not `call-process`, I don't know why this is the case and I agree it is very perculiar but it blocks.

**Comment by lunaryorn** [comment]
Oh sorry I misunderstood you.  I see the problem but I'm unsure whether that's a general issue.

How long does start-process with cargo block on your system?  How does it behave on the terminal?  Try "time emacs --help" versus "time cargo --help".

I wonder why starting cargo should be so much slower than starting Emacs, and I'd like to confirm that it's really a general issue before we use async.

In particular I'd like to see numbers of a start-process cargo approach compared to your async variant.

**Comment by Pireax** [comment]
Here are some numbers:
`time emacs --help`: 0.15s
`time cargo --help`: 6.17s
Note the following numbers are the time it takes to execute `start-process`.
`(start-process "e" nil "emacs" "--help")`: 0.05s
`(start-process "c" nil "cargo" "--help")`: 6.44s

This looks like it may be a cargo issue.

**Comment by lunaryorn** [comment]
@Pireax Wow, that doesn't look right 😳  But I'm not sure whether that's something we should work around in this extension: I've never seen this issue before (with any executable), and I strongly suspect that it's a cargo problem.  After all, it means that cargo takes over 6 seconds to even load the executable into memory and that's really quite unusual.

Could you try to find out whether that's a known issue in Cargo?

**Comment by lunaryorn** [comment]
Besides, please be aware that Flycheck does not claim to support Windows.  Please understand that we are wary to add workarounds for issues that only occur on Windows—an issue would really have to be quite bad and affect a vast majority of Windows users (e.g. https://github.com/flycheck/flycheck/issues/794) for us to explicitly work around it, all the more if the work around comes at the price of an additional dependency who's [interaction with our code](https://github.com/flycheck/flycheck-rust/issues/44#issuecomment-260718917) is not yet clear.

**Comment by Pireax** [comment]
As far as I've been able to find out it isn't a know issue in Cargo. And I understand that this isn't something this package should try to account for. rustc also gives me the same problem, and it gets even wierder, it only blocks when supplying arguments like `--help` without arguments it runs fine, both in the console and in emacs with `start-process`. I also tried other rust programs but didn't find a problem there. This all doesn't make any sense.

Anyways, thank you for taking your time to respond and for creating this package, I'll try to find out what is going on.

**Comment by lunaryorn** [comment]
@Pireax Are you using cygwin or something like that perchance?

**Comment by Pireax** [comment]
@lunaryorn No nothing of the like, just good ol' cmd/windows.

**Comment by lunaryorn** [comment]
@Pireax  You don't happen to use some anti virus software?

I guess you should report that to Rust itself, with detailed information about the system, etc.  It's quite unusual for a program to take 6s to even load, before the process get's spawned.  

If you get the chance I'd recommend to try and reproduce the problem on another Windows system, or at least in a fresh user account with no background programs on the same system.

**Comment by lunaryorn** [comment]
I'll close this issue as wontfix.

**Comment by Pireax** [comment]
@lunaryorn Just booted to safe-mode, no problems there. Turns out it is indeed my anti-virus software😠. Thank's for the help.

**Comment by lunaryorn** [comment]
@Pireax You're welcome :relaxed:

**Comment by ZelphirKaltstahl** [comment]
fwiiw: The long blocking still happens when saving a file, which triggers the flycheck thing to do its thing, at least for me it does. Since saving frequently has become a reflex, it can sometimes be annoying. It would be great if this could be run concurrently.

**Comment by fmdkdd** [comment]
@ZelphirKaltstahl This issue was specifically about `flycheck-rust-find-target` taking a long time, which happens only once when you open a Rust file with the `flycheck-rust` extension.

If you get UI lock ups when saving a file, then that's a different issue.  When saving a file, we do indeed run `cargo rustc`, but we already run it asynchronously with `start-process`.  I don't experience any UI lockup (i.e, I can edit and move the cursor just fine) after saving, even if it may take several seconds to get the diagnostics to come back (on large projects).

**Comment by fmdkdd** [comment]
If you do experience UI locks when saving, then please open a separate issue so we can look into it.

**Comment by ZelphirKaltstahl** [comment]
@fmdkdd Huh, weird, exactly that kind of blocking *is* happening for me. Maybe I am not on the up to date version (afaik I installed from Melpa stable). But I should really make a new issue for this ; )

**Comment by fmdkdd** [comment]
@ZelphirKaltstahl Well, it couldn't hurt to test the latest version.  In fact, we pushed 31 to Melpa stable a few days ago :)
